### PR TITLE
Add support for domain tags in metrics

### PR DESCRIPTION
**Pull request by shreyassrivatsan**
Creating an internal scope structure that allows adding pre-defined tags to metrics. Presently Domain is the one allowed tag.

### CODE REVIEW

File: common/metrics/interfaces.go [hunk 1]
```diff
@@ -43,5 +43,8 @@ type (
 		UpdateGauge(scope int, gauge int, value float64)
 		// Tagged returns a client that adds the given tags to all metrics
 		Tagged(tags map[string]string) Client
+		// Scope return an internal scope that can be used to add additional
+		// information to metrics
+		Scope(scope int) Scope
```
Comment (venkat1109) [inline common/metrics/interfaces.go]: (1) can you get rid of the free style Tagged(tags map[string]string) method in favor of the strongly typed tags ? I don't expect a lot of places we use this today
(2) How about making the signature of Scope as  Scope(scope int, tags ..Tag) ? Any caller that calls Scope() almost always will need a sub-scope with some tags. So, taking the tags as input here saves additional api call / data copy later

Comment (shreyassrivatsan) [reply common/metrics/interfaces.go]: Got rid of tagged. And change the Scope method to take tags. The actual Tagged support is still in the underlying scope returned though.

Comment (venkat1109) [reply common/metrics/interfaces.go]: 👍

File: common/metrics/interfaces.go [hunk 2]
```diff
@@ -39,7 +41,16 @@ type (
...
-		// Tagged returns a client that adds the given tags to all metrics
-		Tagged(tags map[string]string) Client
+		// Scope return an internal scope that can be used to add additional
+		// information to metrics
+		Scope(scope int, tags ...Tag) Scope
+	}
+
+	// Scope is an interface for metrics
+	Scope interface {
+		Counter(id int) tally.Counter
```
[the remark this answers is not in the document]
Comment (shreyassrivatsan) [reply common/metrics/interfaces.go]: Done

File: common/metrics/tags.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	Value() string
+}
+
+// DomainTag wraps a domain tag
+type DomainTag struct {
+	value string
+}
+
+// NewDomainTag returns a new domain tag
+func NewDomainTag(value string) DomainTag {
```
Comment (venkat1109) [inline common/metrics/tags.go]: Just a style preference. If you make DomainTag struct private (domainTag struct) - you could do something like following:

```
client.Scope(metrics.StartWFExecutionScope, 
metrics.DomainTag("domain-name"),
metrics.ShardIDTag(shardID)) 

reads better than 

client.Scope(metrics.StartWFExecutionScope, 
metrics.NewDomainTag("domain-name"),
metrics.NewShardIDTag(shardID))
```

Zap logger uses similar style for typed k/v tags

Comment (shreyassrivatsan) [reply common/metrics/tags.go]: Done.

**Pull request merged**